### DISCUSSION TITLE
App Check keeps returning a invalid token to all devices

### DISCUSSION THREAD

**Discussion by Nayane-19** [firebase/flutterfire#7998] [Q&A] [answered]
I have trouble setting up App-Check on my app(s). The current issue is: On Android, I keep getting the same token "eyJlcnJvciI6IlVOS05PV05fRVJST1IifQ==" on all devices (emulator and physical).
When I decode this token in base64 means: {"error":"UNKNOWN_ERROR"}
I can't find anything in the documentation that helps me with this problem.

Summary: Get same token on all Android devices, and said token still fails with Firebase calls.

**Comment by r-a-o** [comment]
A playstore signed apk will generate proper token. Upload .aab to playstore and download the universal apk from bundle explorer.

**Comment by kapibara91** [comment, marked as answer]
I have the same problem on Android, but when using the **DebugAppCheckProviderFactory** and **debug token**, I can get the normal token.

**Reply by enriquecorpus** [reply]
This is not a solution, this only works on development but not recommended for production releases!

**Comment by chenop** [comment]
Got the same token on release apk for internal testing - any suggestion?

**Reply by chenop** [reply]
After changing the device Im using I manage to get a valid token

**Comment by kNoAPP** [comment]
Shedding some light on the situation from my end...

For each scenario below, I have both the debug and release SHA256 hashes imported to my Firebase console. These were determined by making a debug/release build with my keystores and running `keytool -printcert -jarfile <.aab or .apk>` on the binaries. These have been triple checked, and they are configured correctly.

My app produces the AppCheck token `eyJlcnJvciI6IlVOS05PV05fRVJST1IifQ==` whenever its either:
- Running with `androidProvider` set to `playIntegrity` or `debug` after executing `flutter run --debug` or `flutter run --release`.
- Publishing with `androidProvider` set to `playIntegrity`, `safetyNet`, or `debug` to Firebase App Distribution then downloading/running on my device through Firebase App Distribution.

My app produces a valid AppCheck token whenever its:
- Running with `androidProvider` set to `debug` after executing `flutter run --debug` AND the debug secret produced by the console is imported as a debug key into the Firebase Console.
- Running with `androidProvider` set to deprecated `safetyNet` after executing `flutter run --debug` or `flutter run --release`.

My guess is that the Firebase team has not yet fully fleshed out the new `playIntegrity` attestation method.

**Comment by ElleRyuzaki** [comment]
I'm having the same issue since two days on a Xamarin Android application.
I get the token with this method:
```
public async void GetToken()
{
 var AppCheckToken =  await AppCheck.GetToken(false);

            if (AppCheckToken is AppCheckTokenResult Tok)
            {
                string r;
                r = Tok.Token;
                Preferences.Set("AppCheckToken", r);
            }
}
```
that always return this string `eyJlcnJvciI6IlVOS05PV05fRVJST1IifQ==` on both debug and physical (production) mode. I also tried to make internal tests in Play Console and then download the app from there, but result is always the same.

Here's how I initialize AppCheck:
```
            auth = FirebaseAuth.GetInstance(app);
            PlayIntegrityAppCheckProviderFactory f = new PlayIntegrityAppCheckProviderFactory();
            f.Create(app);
            AppCheck = FirebaseAppCheck.GetInstance(app);
            AppCheck.InstallAppCheckProviderFactory(f);
            AppCheck.SetTokenAutoRefreshEnabled(true);
```

I also checked if all related APIs are active and they are. I don't know what to do

**Reply by chenop** [reply]
Try changing the android you are debugging with

**Reply by ElleRyuzaki** [reply]
I will try and let you know, but in production it must work on the other versions too

**Reply by chenop** [reply]
For sure - my thinking is that maybe after a while of sending requests from the same device it got some how into a black listed and constantly being blocked by Play Integrity.

**Reply by ElleRyuzaki** [reply]
Do you think is an issue related to PlayIntegrity or AppCheck and it will solved with an update?

**Reply by chenop** [reply]
First of all Im not sure Im correct, its just an assumption :-)
If Im correct its an issue in PlayIntegrity - not sure what is the implementation there.

**Reply by ElleRyuzaki** [reply]
Tried now with Android 9 and it returns always the same AppCheck invalid token

**Reply by ElleRyuzaki** [reply]
I noticed that var AppCheckToken is always null and if I debug and see step-to-step run it throws this error:
`
  --- End of managed Firebase.FirebaseException stack trace ---
com.google.firebase.FirebaseException: null
	at com.google.firebase.appcheck.internal.DefaultFirebaseAppCheck.lambda$getToken$1(DefaultFirebaseAppCheck.java:205)
	at com.google.firebase.appcheck.internal.DefaultFirebaseAppCheck$$ExternalSyntheticLambda1.then(Unknown Source:0)
	at com.google.android.gms.tasks.zze.run(com.google.android.gms:play-services-tasks@@18.0.2:1)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7892)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
Caused by: java.lang.NumberFormatException: null
	at java.lang.Long.parseLong(Long.java:699)
	at java.lang.Long.parseLong(Long.java:861)
	at com.google.firebase.appcheck.playintegrity.internal.PlayIntegrityAppCheckProvider.lambda$getPlayIntegrityAttestation$4$com-google-firebase-appcheck-playintegrity-internal-PlayIntegrityAppCheckProvider(PlayIntegrityAppCheckProvider.java:105)
	at com.google.firebase.appcheck.playintegrity.internal.PlayIntegrityAppCheckProvider$$ExternalSyntheticLambda1.then(Unknown Source:4)
	at com.google.android.gms.tasks.zzo.run(com.google.android.gms:play-services-tasks@@18.0.2:1)
	... 8 more
`

**Reply by chenop** [reply]
@ElleRyuzaki Do you know for sure that Firebase AppCheck support Play Integrity?
Im asking since Im using react-native-firebase and it turn out that [this specific library does not support PlayIntegrity ](https://github.com/invertase/react-native-firebase/issues/6753#issuecomment-1336475520) yet.

**Reply by ElleRyuzaki** [reply]
I get same problems with SafetyNetProviderFactory 😭

**Reply by mohamedbrahimi** [reply]
I have the same issue token not valid! Does any one have a solution for this?

**Reply by ElleRyuzaki** [reply]
Unfortunately I didn't find a solution.
Firebase support told me that Firebase is not officially supported in Xamarin, by the way in their opinion there's a problem with objects initialization, which are always null.

**Comment by Andrushka1012** [comment]
any news here? Today 16.05.23 and still no solution? 😢

**Comment by jerry7991** [comment]
Hi everyone, I am getting the below error can anyone help, please?
`E/AppCheck: Token retrieval failed
    com.google.firebase.FirebaseException: Error returned from API. code: 403 body: App attestation failed.`

**Comment by dehypnosis** [comment]
Idiots are sending deprecation warning without fixing such issues.

**Comment by Benjith-Zoondia** [comment]
Still same issue

**Comment by Mohit-Hasan** [comment]
No proper docs or solution about this matter, shame on google!
